consensus/istanbul/core/backlog.go

```diff
@@ -103,114 +116,181 @@ func (c *core) storeBacklog(msg *istanbul.Message, src istanbul.Validator) {
+	backlogSeqs := make([]uint64, len(c.backlogBySeq))
+	i := 0
+	for k := range c.backlogBySeq {
+		backlogSeqs[i] = k
+		i++
+	}
+	sort.Slice(backlogSeqs, func(i, j int) bool {
+		return backlogSeqs[i] < backlogSeqs[j]
+	})
+	return backlogSeqs
+}
+
+// Drain a backlog for a given sequence, passing each to optional callback.
+// Call with backlogsMu held.
+func (c *core) drainBacklogForSeq(seq uint64, cb func(*istanbul.Message, istanbul.Validator)) {
+	backlogForSeq := c.backlogBySeq[seq]
+	if backlogForSeq == nil {
+		return
+	}
+
+	for !backlogForSeq.Empty() {
+		m := backlogForSeq.PopItem()
+		msg := m.(*istanbul.Message)
+		if cb != nil {
+			_, src := c.valSet.GetByAddress(msg.Address)
+			if src != nil {
+				cb(msg, src)
+			}
+		}
+		c.backlogCountByVal[msg.Address]--
+		c.backlogTotal--
+	}
+	delete(c.backlogBySeq, seq)
 }
 
 func (c *core) processBacklog() {
+
 	c.backlogsMu.Lock()
 	defer c.backlogsMu.Unlock()
 
-	for src, backlog := range c.backlogs {
-		if backlog == nil {
-			continue
-		}
+	for _, seq := range c.getSortedBacklogSeqs() {
 
-		logger := c.logger.New("from", src, "state", c.state, "cur_round", c.current.Round(), "cur_seq", c.current.Sequence(), "func", "processBacklog")
-		isFuture := false
-
-		// We stop processing if
-		//   1. backlog is empty
-		//   2. The first message in queue is a future message
-		for !(backlog.Empty() || isFuture) {
-			m, prio := backlog.Pop()
-			msg := m.(*istanbul.Message)
-			var view *istanbul.View
-			switch msg.Code {
-			case istanbul.MsgPreprepare:
-				var m *istanbul.Preprepare
-				err := msg.Decode(&m)
-				if err == nil {
-					view = m.View
-				}
-			case istanbul.MsgPrepare:
-				fallthrough
-			case istanbul.MsgCommit:
-				var sub *istanbul.Subject
-				err := msg.Decode(&sub)
-				if err == nil {
-					view = sub.View
+		logger := c.logger.New("state", c.state, "seq", seq)
+
+		if seq < c.currentView().Sequence.Uint64() {
+			// Earlier sequence. Prune all messages.
+			c.drainBacklogForSeq(seq, nil)
+		} else if seq == c.currentView().Sequence.Uint64() {
+			// Current sequence. Process all in order.
+			c.drainBacklogForSeq(seq, func(msg *istanbul.Message, src istanbul.Validator) {
+				var view *istanbul.View
+				switch msg.Code {
+				case istanbul.MsgPreprepare:
+					var m *istanbul.Preprepare
+					err := msg.Decode(&m)
+					if err == nil {
+						view = m.View
+					}
+				case istanbul.MsgPrepare:
+					fallthrough
+				case istanbul.MsgCommit:
+					var sub *istanbul.Subject
+					err := msg.Decode(&sub)
+					if err == nil {
+						view = sub.View
+					}
+				case istanbul.MsgRoundChange:
+					var rc *istanbul.RoundChange
+					err := msg.Decode(&rc)
+					if err == nil {
+						view = rc.View
+					}
 				}
-			case istanbul.MsgRoundChange:
-				var rc *istanbul.RoundChange
-				err := msg.Decode(&rc)
-				if err == nil {
-					view = rc.View
+				if view == nil {
+					logger.Debug("Nil view", "msg", msg)
```

Comment:
Should this not be a more worrying level given that a message in this sequence backlog somehow we can extract the view of?

Reply:
Changed to `Error`